transport-classes-io_uring/src/main/java/io/netty5/incubator/channel/uring/IOUringDatagramChannel.java

```diff
@@ -0,0 +1,535 @@
+
+        // Since we remove the messages from WriteSink/OutboundBuffer, it falls to us to close the buffer, and complete
+        // the promise.
+        promise.asFuture().addListener(data, CLOSE_BUFFER);
+        short pendingId = pendingWrites.addPending(promise);
+
+        try (var itr = data.forEachComponent()) {
+            var cmp = itr.firstReadable();
+            assert cmp != null;
+            int fd = fd().intValue();
+            if (connected) {
+                submissionQueue.addSend(fd, cmp.readableNativeAddress(), 0, cmp.readableBytes(), pendingId);
+            } else {
+                CachedMsgHdrMemory msgHdr = nextMsgHdr();
+                msgHdr.write(socket, remoteAddress, cmp.readableNativeAddress(), cmp.readableBytes(), (short) 0);
+                submissionQueue.addSendmsg(fd, msgHdr.address(), 0, pendingId);
+                promise.asFuture().addListener(msgHdr);
+            }
+        }
+        return writeHandle().lastWrite(data.readableBytes(), data.readableBytes(), 1);
+    }
+
+    @NotNull
+    private CachedMsgHdrMemory nextMsgHdr() {
+        CachedMsgHdrMemory msgHdr = msgHdrCache.pollFirst();
+        if (msgHdr == null) {
+            msgHdr = new CachedMsgHdrMemory(msgHdrCache);
+        }
+        return msgHdr;
+    }
+
+    @Override
+    void writeComplete(int result, long udata) {
+        Promise<Void> promise = pendingWrites.removePending(UserData.decodeData(udata));
+        if (result < 0) {
+            promise.setFailure(Errors.newIOException("send/sendmsg", result));
+        } else {
+            promise.setSuccess(null);
+        }
+    }
+
+    @Override
+    protected void doDisconnect() throws Exception {
+        super.doDisconnect();
+        connected = false;
+        cacheAddresses(local, null);
+        remote = null;
+    }
+
+    @Override
+    protected boolean doConnect(SocketAddress remoteAddress, SocketAddress localAddress, Buffer initialData) throws Exception {
+        boolean connected = super.doConnect(remoteAddress, localAddress, initialData);
+        if (connected) {
+            this.connected = true;
+        }
+        return connected;
+    }
+
+    @Override
+    protected void doShutdown(ChannelShutdownDirection direction) throws Exception {
+        switch (direction) {
+            case Inbound:
+                inputShutdown = true;
+                break;
+            case Outbound:
+                outputShutdown = true;
+                break;
+            default:
+                throw new IllegalArgumentException("Unknown direction: " + direction);
+        }
+    }
+
+    @Override
+    public boolean isShutdown(ChannelShutdownDirection direction) {
+        if (!isActive()) {
+            return true;
+        }
+        switch (direction) {
+            case Inbound:
+                return inputShutdown;
+            case Outbound:
+                return outputShutdown;
+            default:
+                throw new AssertionError();
```

Comment:
nit: Either use the `AssertationError` everywhere or ` throw new IllegalArgumentException("Unknown direction: " + direction);`

Reply:
Looks like `AssertionError` is the most common.